google/cloud/forseti/scanner/audit/retention_rules_engine.py

```diff
@@ -0,0 +1,318 @@
+    """
+
+    rttRuleViolation = namedtuple(
+        'RuleViolation',
+        ['resource_name', 'resource_type', 'full_name', 'rule_name',
+         'rule_index', 'violation_type', 'violation_describe'])
+
+    def __init__(self, rule_name, rule_index, min_retention, max_retention):
+        """Initialize.
+
+        Args:
+            rule_name (str): Name of the loaded rule
+            rule_index (int): The index of the rule
+            min_retention(int): minimum value of the age in lifecycle
+            max_retention(int): maximum value of the age in lifecycle
+        """
+        self.rule_name = rule_name
+        self.rule_index = rule_index
+        self.min_retention = min_retention
+        self.max_retention = max_retention
+
+    def generate_rule_violation(self, buckets_lifecycle, describe):
+        """generate a violation.
+
+        Args:
+            buckets_lifecycle (RetentionBucket): The info of the bucket
+            describe (str): The description of the violation
+        Returns:
+            rttRuleViolation: The violation
+        """
+
+        return self.rttRuleViolation(
+            resource_name=buckets_lifecycle.name,
+            resource_type=buckets_lifecycle.type,
+            full_name=buckets_lifecycle.full_name,
+            rule_name=self.rule_name,
+            rule_index=self.rule_index,
+            violation_type=VIOLATION_TYPE,
+            violation_describe=describe
+        )
+
+    # TODO: The naming is confusing and needs to be fixed in all scanners.
+    def find_buckets_violations(self, buckets_lifecycle):
+        """Get a generator for violations
+
+        Args:
+            buckets_lifecycle (RetentionBucket): The info of the bucket
+        Yields:
+            rttRuleViolation: All violations of the bucket breaking the rule.
+        """
+        minretention = self.min_retention
+        maxretention = self.max_retention
+        exist_match = False
+        for lci in buckets_lifecycle.lifecycleitems:
+            age = lci.get('condition', {}).get('age', None)
+            if age is None:
+                continue
+            if(minretention != None and age < minretention):
+                yield self.generate_rule_violation(
+                    buckets_lifecycle, 'age %d is smaller than \
+the minimum retention %d' % (age, minretention))
+                continue
+            if(maxretention != None and age > maxretention):
+                yield self.generate_rule_violation(
+                    buckets_lifecycle, 'age %d is larger than \
+the maximum retention %d' % (age, maxretention))
+                continue
+            if 'createdBefore' in lci.get('condition', {}):
+                continue
+            if 'matchesStorageClass' in lci.get('condition', {}):
+                continue
+            if 'numNewerVersions' in lci.get('condition', {}):
+                continue
+            if 'isLive' in lci.get('condition', {}):
+                continue
```

Comment:
Move this to a helper function that works out the actual retention age (could be None if there is no delete rule). Then in this function can check that min_retention <= actual_retention <= max_retention